### ISSUE TITLE
Sample app keeps sending data after StopCapture directive is received

### ISSUE THREAD

**Issue by ryter--** [alexa/alexa-avs-sample-app#893]
Hi,

I am working with the Conexant2Mic Raspberry Pi kit together with the current AVS Sample App and am experiencing significant latency problems (~10s average from the end of user speech until the beginning of Alexa's speech response).
At first, I thought that those were due to me not streaming audio data or waiting until all of the sound data from Alexa was received. However, after reviewing the timing of events in the application log, the timing of out- and incoming packets between my client and AVS, and the time of entry into my lambda function, I think I found out what actually happens:
After the StopCapture directive is received by the client, the client keeps on sending packets to AVS (sometimes for more than 8s) until it receives the 200 response code and finally stops. The lambda function is called by Alexa right after the StopDirective is sent, but the speech output is delayed until the initial request from the client is finished. Testing on the Echo Dot shows minimal latency (1-2 s), so the problem should not originate from the lambda function.

I did not change anything in the code and am currently digging through the calling hierarchy to find out what is happening. (E.g. does microphone.stopCapture() not cause the request to terminate?)

Am I missing or misunderstanding something?
What is the best way to actually stop the request once the StopCapture directive is received?

Thanks in advance

**Comment by AlphaDrakonix999** [comment]
Hello Ryter,
Thank you for reaching out to us with your inquiry.  I would suggest to check if:
1 - Does your client immediately close the microphone and stop listening for the user's speech?
2 - Are you using a 'profile' in your Recognize event that support cloud-endpointing, such as NEAR_FIELD or FAR_FIELD?
Source of information: https://developer.amazon.com/public/solutions/alexa/alexa-voice-service/reference/speechrecognizer#stopcapture
Please let us know if this information was of any help to you.
Best,

**Comment by ryter--** [comment]
Hello Drakonix,

1 - yes. I did not change the javaclient, and as far as I can see, once the AVSController sees a StopCapture, after a few redirections its method stopRecording() is called, wherein microphone.stopCapture() should close the microphone line.
2 - yes, I tried both NEAR_FIELD (which was selected as SpeechProfile form the start) as well as FAR_FIELD. Both yield similar results.

My best guess is that after the microphone is closed the client still has some cached data which it has not yet sent and continues to send it until all of the recorded speech is transmitted. This adds several seconds to the overall latency.
To illustrate, here is an excerpt from the application log:

_08:22:35.078 [DownchannelRequestThread] INFO  com.amazon.alexa.avs.http.MessageParser - Response metadata: 
{
  "directive" : {
    "header" : {
      "namespace" : "SpeechRecognizer",
      "name" : "StopCapture",
      "messageId" : "52e80185-bc56-47f8-b567-31f1ae1221dd"
    },
    "payload" : { }
  }
}
08:22:35.081 [IndependentDirectiveThread] INFO  com.amazon.alexa.avs.AVSController - Handling directive: SpeechRecognizer.StopCapture
08:22:35.082 [IndependentDirectiveThread] DEBUG com.amazon.alexa.avs.SpeechRequestAudioPlayerPauseController - Finished listening to user speech
08:22:35.083 [IndependentDirectiveThread] DEBUG com.amazon.alexa.avs.wakeword.WakeWordIPCSocket - Sending command IPC_RESUME_WAKE_WORD_ENGINE to all connected clients
08:22:40.730 [RequestThread] INFO  com.amazon.alexa.avs.http.AVSClient - Response code: 200_

Going through the tcpdump output, I found the packet with the StopCapture directive:
_08:22:35.075803 IP 54.239.38.218.https > raspberrypi.59556: Flags [P.], seq 253:516, ack 115426, win 948, length 263_

After this packet is received by the client, it continues to send packets for a couple of seconds, then stops and waits for the ACKs from alexa until finally the response code 200 is received. Then, the speech response from Alexa is transmitted. 

It seems to me that by cutting short the request transmitting the captured sound directly after the receipt of the StopCapture directive, the overall latency should drop significantly.

But again my question: Am I missing something? Is there another way to reduce the latency?
